### PR TITLE
Go implementation of Heap Sort #2147

### LINKED ISSUE #2147: Heap Sort
- [x] C
- [x] C++
- [x] C#
- [ ] CoffeeScript
- [x] Dart
- [x] Go - @Shr03mink
- [x] Java
- [x] JS
- [ ] Kotlin
- [x] Php - @anmolkaur18
- [x] Python
- [x] Readme
- [x] Ruby
- [ ] Rust
- [ ] TypeScript

### PR DESCRIPTION
**Pull request by Shr03mink**
Fixes  #2147 

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
-  Implemented heap sort by creating Max heap
-  

#### Languages Used:
-  Go
- 

#### Files Added:
- heap_sort.go
- 

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by Shr03mink** [comment]
I have implemented Heap Sort in Go. Please review and merge. Thank you.

File: Heap_Sort/heap_sort.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+5
+6
+7
+
+Original Array: [0 12 11 13 5 6 7]
+
+Sorted array is
+0 5 6 7 11 12 13 
+
+*/
```
Comment (singh-shreya6) [inline Heap_Sort/heap_sort.go:98]: Please add a blank line at end of file

File: Heap_Sort/heap_sort.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+func main() {
+	var n int
+	fmt.Println("Enter the number of elements :") 
+	fmt.Scan(&n) //accepting the number of elements from the user
+	arr := make([]int, n) //creating an array of n elements
+	
+	for i := 0; i < n; i++ {
+      fmt.Scan(&arr[i])  //accepting input elements
+    }
```
Comment (singh-shreya6) [inline Heap_Sort/heap_sort.go:73]: Incorrect indentation, please keep 4 spaces.

File: Heap_Sort/heap_sort.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	fmt.Println(arr)
+    
+    fmt.Println("Sorted array is") 
+	heapSort(arr, n)
```
Comment (singh-shreya6) [inline Heap_Sort/heap_sort.go:77]: Incorrect indentation.

File: Heap_Sort/heap_sort.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        max_heapify(arr, i, 0); 
+    } 
+}
+ 
+
+func printArray(arr []int, no int) {  
+   
+    //printing the sorted array
+    for i := 0; i <= 10; i++ {
+        fmt.Printf("\n%d",arr[i])
```
Comment (singh-shreya6) [inline Heap_Sort/heap_sort.go:59]: Please add space after comma

File: Heap_Sort/heap_sort.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        // call heapify on the reduced heap 
+        max_heapify(arr, i, 0); 
+    } 
+}
+ 
+
+func printArray(arr []int, no int) {  
+   
+    //printing the sorted array
+    for i := 0; i <= 10; i++ {
```
Comment (singh-shreya6) [inline Heap_Sort/heap_sort.go:58]: why running the loop till 10. Please use the value of n you have taken as input

File: Heap_Sort/heap_sort.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    // Build heap (rearrange array) 
+    for  i := n / 2 - 1; i >= 0; i-- { 
+        max_heapify(arr, n, i)
+    }
+  
+    // One by one extract an element from heap 
+    for  i := n - 1; i >= 0; i-- { 
+        // Move current root to end 
+        arr[0], arr[i]=arr[i],arr[0];  //swapping
```
Comment (singh-shreya6) [inline Heap_Sort/heap_sort.go:47]: Please add space around = and one space after comma

File: Heap_Sort/heap_sort.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
+
+/*	Heap Sort is a comparison-based sorting method performed on a Heap data structure.
+    Here we build a max heap first to later sort the elements in the heap in ascending order.
```
Comment (singh-shreya6) [inline Heap_Sort/heap_sort.go:3]: Please indent Here to fall in line with Heap

**Comment by Shr03mink** [comment]
i have made all the changes. Thank you!

**Comment by Shr03mink** [comment]
Is it alright now?

**Comment by Shr03mink** [comment]
HI. I am actually having trouble understanding how to squash the commits. Could you please help me with it.

**Comment by somya-kapoor** [comment]
For reference http://blog.steveklabnik.com/posts/2012-11-08-how-to-squash-commits-in-a-github-pull-request

**Comment by Shr03mink** [comment]
i am not able to fetch the right branch. Whenever i fetch the origin branch and then try rebase it shows me the commits of other contributors of the issue as well as my original commit.

**Comment by singh-shreya6** [comment]
As shared by one of the participants, please follow this procedure to squash commits:
Suppose the name of my working branch is my_branch

I will switch to my branch
git checkout my_branch

After that, I will do all my work in that branch. I will modify files, work on them and finally push all the commits and changes.

Now I need to squash the commits [right now I am in my_branch]
git rebase -i upstream/master

A screen will open up showing the list of all commits with different labels along with it.
Keep the first commit as pick
& change all other commits to squash

This will squash (merge) the respective commit onto previous one
After changing all labels to squash except the first one, save & exit.

Next, a screen will appear showing all commit messages, If you want, you can change the commit messages or leave them as it is. Save and exit.

This will rebase all the commits in case there are no merge conflicts.

Last step is to force push the commits so that commit count becomes one.
git push --force origin my_branch

Done.

**Comment by singh-shreya6** [comment]
Are you working on it?

**Comment by Shr03mink** [comment]
yes, i am. Was having an assignment submission. ll be done with this by tomorrow.

**Comment by Shr03mink** [comment]
I had intially not saved any changes locally. I now understood the working and made the above changes locally and then i squashed my 2 commits. Do i now create a new pull request for the squashed commits?

**Comment by Shr03mink** [comment]
Hi. I had unknowingly created 2 branches and i had made all the changes locally in the 2nd branch instead of this branch from where i created the pull request. So, i made a new pull request with squashed commits, following the steps you mentioned above. I am really sorry for causing you all this trouble but could you please refer to the new PR instead of this one. The PR reference is #2884

**Pull request closed without merging**